Bugsnag/KSCrash/Source/KSCrash/Recording/Tools/BSG_KSMachHeaders.c

```diff
@@ -195,43 +189,42 @@ bool bsg_mach_headers_populate_info(const struct mach_header *header, intptr_t s
     info->name = imageName;
     info->slide = slide;
     info->unloaded = FALSE;
-    info->next = NULL;
+    atomic_store(&info->next, NULL);
     
     return true;
 }
 
-void bsg_mach_headers_add_image(const struct mach_header *header, intptr_t slide) {
+static void add_image(const struct mach_header *header, intptr_t slide) {
     BSG_Mach_Header_Info *newImage = calloc(1, sizeof(BSG_Mach_Header_Info));
-    if (newImage != NULL) {
-        if (bsg_mach_headers_populate_info(header, slide, newImage)) {
-            dispatch_sync(bsg_g_serial_queue, ^{
-                if (bsg_g_mach_headers_images_head == NULL) {
-                    bsg_g_mach_headers_images_head = newImage;
-                } else {
-                    bsg_g_mach_headers_images_tail->next = newImage;
-                }
-                bsg_g_mach_headers_images_tail = newImage;
-                if (header == &__dso_handle) {
-                    g_self_image = newImage;
-                }
-            });
-        } else {
-            free(newImage);
-        }
+    if (newImage == NULL) {
+        return;
+    }
+
+    if (!bsg_mach_headers_populate_info(header, slide, newImage)) {
+        free(newImage);
+        return;
+    }
+
+    BSG_Mach_Header_Info *oldTail = atomic_load(&g_images_tail);
+    atomic_store(&g_images_tail, newImage);
+    atomic_store(&oldTail->next, newImage);
```

Comment:
I think we're missing a CAS operation here to make sure the load/store don't race?

Something like:

```suggestion
    BSG_Mach_Header_Info *oldTail = atomic_load(&g_images_tail);
    for(;;) {
      if(atomic_compare_exchange_strong(&g_images_tail, &oldTail, newImage)) {
        break;
      }
    }
    }
    atomic_store(&oldTail->next, newImage);
```